mcs/class/Mono.Posix/Mono.Unix/UnixMarshal.cs

```diff
@@ -325,27 +325,34 @@ public static IntPtr StringToHeap (string s, int index, int count, Encoding enco
 			if (encoding == null)
 				throw new ArgumentNullException ("encoding");
 
-			int min_byte_count = encoding.GetMaxByteCount(1);
-			char[] copy = s.ToCharArray (index, count);
-			byte[] marshal = new byte [encoding.GetByteCount (copy) + min_byte_count];
+			if (index < 0 || count < 0)
```

Comment:
This should also ensure that `(index + count) < s.Length`, as we don't want to permit reading random memory, e.g. `UnixMarshal.STringToHeap ("1", 1, 2)` should throw.

For example, see the [`Encoding.GetBytes(string, int, int, byte[], int)` documentation](https://msdn.microsoft.com/en-us/library/c20ss51h(v=vs.110).aspx):

> `ArgumentOutOfRangeException`: 
> *charIndex* or *charCount* or *byteIndex* is less than zero.
> -or-
> *charIndex* and *charCount* do not denote a valid range in chars.
> -or-
> *byteIndex* is not a valid index in bytes.